File: backend/auth/tools/hasher.go
package tools

import "golang.org/x/crypto/bcrypt"

var (
	hash_salt = 10
)

type Hasher struct{}

func NewHasher() *Hasher {
	return new(Hasher)
}

func (h *Hasher) Hash(s string) (string, error) {
	hashed_string_bytes, err := bcrypt.GenerateFromPassword([]byte(s), hash_salt)
	return string(hashed_string_bytes), err
}

func (h *Hasher) Compare(s, hashed_s string) bool {
	err := bcrypt.CompareHashAndPassword([]byte(hashed_s), []byte(s))
	return err == nil
}

File: apps/auth/src/utils/hasher.rs
pub struct Hasher;

const HASH_SALT: u32 = 10;

impl Hasher {
  pub fn hash<S: AsRef<[u8]>>(s: S) -> Result<String, ()> {
    Ok(bcrypt::hash(s, HASH_SALT).map_err(|err| {
      eprintln!("Error hashing password: {}", err);
    })?)
  }

  pub fn verify<S: AsRef<[u8]>>(s: S, hash: &str) -> Result<bool, ()> {
    Ok(bcrypt::verify(s, hash).map_err(|err| {
      eprintln!("Error verifying password: {}", err);
    })?)
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_hasher() {
    let s = "some_string_9489";
    let hash = Hasher::hash(s).unwrap();
    assert_eq!(true, hash.len() > 1);
    assert_eq!(true, Hasher::verify(s, &hash).unwrap());
  }
}
